File: Remote/Source/Remote Elements/REView/ButtonGroupView.m
//
// ButtonGroupView.m
// Remote
//
// Created by Jason Cardwell on 5/24/11.
// Copyright 2011 Moondeer Studios. All rights reserved.
//

#import "RemoteElementView_Private.h"
#import "ButtonGroup.h"
#import "Button.h"

static int ddLogLevel   = LOG_LEVEL_DEBUG;
static int msLogContext = (LOG_CONTEXT_REMOTE | LOG_CONTEXT_FILE | LOG_CONTEXT_CONSOLE);
#pragma unused(ddLogLevel,msLogContext)


@interface ButtonGroupView ()

@property (nonatomic, weak) NSLayoutConstraint       * tuckedConstraint;
@property (nonatomic, weak) NSLayoutConstraint       * untuckedConstraint;
@property (nonatomic, weak) MSSwipeGestureRecognizer * tuckGesture;
@property (nonatomic, weak) MSSwipeGestureRecognizer * untuckGesture;

@property (nonatomic, assign)  UISwipeGestureRecognizerDirection tuckDirection;
@property (nonatomic, assign)  UISwipeGestureRecognizerDirection untuckDirection;
@property (nonatomic, assign)  MSSwipeGestureRecognizerQuadrant  quadrant;

@end

@implementation ButtonGroupView


- (void)updateConstraints {

  [super updateConstraints];

  NSString * nametag = ClassNametagWithSuffix(@"Label");

  if (self.label && ![self constraintsWithNametagPrefix:nametag]) {
    NSString * constraints =
      $(@"'%1$@' label.width = self.width\n"
        "'%1$@' label.height = self.height\n"
        "'%1$@' label.centerX = self.centerX\n"
        "'%1$@' label.centerY = self.centerY", nametag);

    [self addConstraints: [NSLayoutConstraint constraintsByParsingString:constraints
                                                                   views:@{@"self": self, @"label": self.label}]];
  }


}

- (CGSize)intrinsicContentSize {
  switch ((uint8_t)self.model.role) {
    case REButtonGroupRoleToolbar: return CGSizeMake(MainScreen.bounds.size.width, 44.0);
    default:                       return CGSizeMake(UIViewNoIntrinsicMetric, UIViewNoIntrinsicMetric);
  }
}

- (void)buttonViewDidExecute:(ButtonView *)buttonView {
  if (self.model.autohide) [self performSelector:@selector(tuck) withObject:nil afterDelay:1.0];
}

- (void)tuck {
  if (self.model.isPanel && self.tuckedConstraint && self.untuckedConstraint) {
    [UIView animateWithDuration:0.25
                     animations:^{
                       self.untuckedConstraint.priority = 1;
                       self.tuckedConstraint.priority = 999;
                       [self.window setNeedsUpdateConstraints];
                       [self setNeedsLayout];
                       [self layoutIfNeeded];
                     }

                     completion:^(BOOL finished) {
                       _tuckGesture.enabled = NO;
                       _untuckGesture.enabled = YES;
                     }];
  }
}

- (void)untuck {
  if (self.model.isPanel && self.tuckedConstraint && self.untuckedConstraint) {
    [UIView animateWithDuration:0.25
                     animations:^{
                       self.tuckedConstraint.priority = 1;
                       self.untuckedConstraint.priority = 999;
                       [self.window setNeedsUpdateConstraints];
                       [self setNeedsLayout];
                       [self layoutIfNeeded];
                     }

                     completion:^(BOOL finished) {
                       _tuckGesture.enabled = YES;
                       _untuckGesture.enabled = NO;
                     }];

  }
}

- (void)handleSwipe:(UISwipeGestureRecognizer *)gesture {
  if (gesture.state == UIGestureRecognizerStateRecognized) {
    if (gesture == _tuckGesture)
      [self tuck];
    else if (gesture == _untuckGesture)
      [self untuck];
  }
}

- (void)attachTuckGestures {
  MSSwipeGestureRecognizer * tuck = [[MSSwipeGestureRecognizer alloc] initWithTarget:self
                                                                              action:@selector(handleSwipe:)];
  tuck.nametag   = $(@"'%@'-tuck", self.name);
  tuck.enabled   = NO;
  tuck.direction = self.tuckDirection;
  tuck.quadrant  = self.quadrant;
  [self.window addGestureRecognizer:tuck];
  self.tuckGesture = tuck;

  MSSwipeGestureRecognizer * untuck = [[MSSwipeGestureRecognizer alloc] initWithTarget:self
                                                                                action:@selector(handleSwipe:)];
  untuck.nametag   = $(@"'%@'-untuck", self.name);
  untuck.direction = self.untuckDirection;
  untuck.quadrant  = self.quadrant;
  [self.window addGestureRecognizer:untuck];
  self.untuckGesture = untuck;
}

- (void)didMoveToWindow { if (self.model.isPanel && !self.isEditing && self.window) [self attachTuckGestures]; }

- (MSDictionary *)kvoRegistration {
  // TODO: observe panel location

  MSDictionary * reg = [super kvoRegistration];
  reg[@"label"] = ^(MSKVOReceptionist * receptionist) {
      id newValue = receptionist.change[NSKeyValueChangeNewKey];
      _label.attributedText = (ValueIsNotNil(newValue) ? (NSAttributedString *)newValue : nil);
    };

  return reg;
}

- (void)initializeIVARs {
  [super initializeIVARs];

  self.shrinkwrap  = YES;
  self.resizable   = YES;
  self.moveable    = YES;
  self.cornerRadii = CGSizeMake(5.0f, 5.0f);

  if ((self.model.role & REButtonGroupRoleToolbar) == REButtonGroupRoleToolbar) {
    [self setContentCompressionResistancePriority:UILayoutPriorityRequired
                                          forAxis:UILayoutConstraintAxisHorizontal];
    [self setContentCompressionResistancePriority:UILayoutPriorityRequired
                                          forAxis:UILayoutConstraintAxisVertical];
  }
}

- (void)didMoveToSuperview {
  [super didMoveToSuperview];

  if (self.superview && [self.model isPanel] && !self.isEditing) {

    NSLayoutAttribute attribute1, attribute2;

    switch (self.model.panelLocation) {
      case REPanelLocationTop:
        attribute1           = NSLayoutAttributeBottom;
        attribute2           = NSLayoutAttributeTop;
        self.tuckDirection   = UISwipeGestureRecognizerDirectionUp;
        self.untuckDirection = UISwipeGestureRecognizerDirectionDown;
        self.quadrant        = MSSwipeGestureRecognizerQuadrantUp;
        break;

      case REPanelLocationBottom:
        attribute1           = NSLayoutAttributeTop;
        attribute2           = NSLayoutAttributeBottom;
        self.tuckDirection   = UISwipeGestureRecognizerDirectionDown;
        self.untuckDirection = UISwipeGestureRecognizerDirectionUp;
        self.quadrant        = MSSwipeGestureRecognizerQuadrantDown;
        break;

      case REPanelLocationLeft:
        attribute1           = NSLayoutAttributeRight;
        attribute2           = NSLayoutAttributeLeft;
        self.tuckDirection   = UISwipeGestureRecognizerDirectionLeft;
        self.untuckDirection = UISwipeGestureRecognizerDirectionRight;
        self.quadrant        = MSSwipeGestureRecognizerQuadrantLeft;
        break;

      case REPanelLocationRight:
        attribute1           = NSLayoutAttributeLeft;
        attribute2           = NSLayoutAttributeRight;
        self.tuckDirection   = UISwipeGestureRecognizerDirectionRight;
        self.untuckDirection = UISwipeGestureRecognizerDirectionLeft;
        self.quadrant        = MSSwipeGestureRecognizerQuadrantRight;
        break;

      default:
        assert(NO);
        break;
    }

    self.tuckedConstraint = [NSLayoutConstraint constraintWithItem:self
                                                     attribute:attribute1
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.superview
                                                     attribute:attribute2
                                                    multiplier:1.0f
                                                      constant:0.0f];
    self.tuckedConstraint.priority = 999;
    self.tuckedConstraint.nametag  = REButtonGroupPanelNametag;
    self.untuckedConstraint        = [NSLayoutConstraint constraintWithItem:self
                                                              attribute:attribute2
                                                              relatedBy:NSLayoutRelationEqual
                                                                 toItem:self.superview
                                                              attribute:attribute2
                                                             multiplier:1.0f
                                                               constant:0.0f];
    self.untuckedConstraint.priority = 1;
    self.untuckedConstraint.nametag  = REButtonGroupPanelNametag;
    [self.superview addConstraints:@[self.tuckedConstraint, self.untuckedConstraint]];
  }
}

- (void)addSubelementView:(ButtonView *)view {
  if (self.locked) [view setValuesForKeysWithDictionary:@{ @"resizable" : @NO,
                                                           @"moveable"  : @NO }];

  if (view.model.role == REButtonRoleTuck) {
    __weak ButtonGroupView * weakself = self;
    view.tapAction = ^{ [weakself tuck]; };
  }

  [super addSubelementView:view];
}

- (void)addInternalSubviews {
  [super addInternalSubviews];

  UILabel * label = [UILabel newForAutolayout];
  label.backgroundColor = ClearColor;
  [self addViewToContent:_label];
  self.label = label;
}

- (void)setEditingMode:(REEditingMode)mode {
  [super setEditingMode:mode];

  self.resizable = (self.editingMode == REEditingModeNotEditing) ? YES : NO;
  self.moveable  = (self.editingMode == REEditingModeNotEditing) ? YES : NO;

  if (self.editingMode == RERemoteEditingMode)
    self.subelementInteractionEnabled = NO;

  else if (self.editingMode == REButtonGroupEditingMode)
    self.subelementInteractionEnabled = YES;

}

@end

File: Remote/Source/Remote Elements/Views/ButtonGroupView.swift
//
//  ButtonGroupView.swift
//  Remote
//
//  Created by Jason Cardwell on 11/07/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import UIKit
import MoonKit

class ButtonGroupView: RemoteElementView {

	var label: UILabel!

	weak var tuckedConstraint: NSLayoutConstraint?
	weak var untuckedConstraint: NSLayoutConstraint?
	weak var tuckGesture: MSSwipeGestureRecognizer?
	weak var untuckGesture: MSSwipeGestureRecognizer?



	/** tuck */
	func tuck() {
		if model.isPanel && tuckedConstraint != nil && untuckedConstraint != nil {
			UIView.animateWithDuration(0.25, animations: {
				self.untuckedConstraint.priority = 1
				self.tuckedConstraint.priority = 999
				self.window.setNeedsUpdateConstraints()
				self.setNeedsLayout()
				self.layoutIfNeeded()
				},
				completion: {
					(finished: Bool) -> Void in
						self.tuckGesture.enabled = false
						self.untuckGesture.enabled = true
				})
		}
	}

	/** untuck */
	func untuck() {
		if model.isPanel && tuckedConstraint != nil && untuckedConstraint != nil {
			UIView.animateWithDuration(0.25, animations: {
				self.untuckedConstraint.priority = 999
				self.tuckedConstraint.priority = 1
				self.window.setNeedsUpdateConstraints()
				self.setNeedsLayout()
				self.layoutIfNeeded()
				},
				completion: {
					(finished: Bool) -> Void in
						self.tuckGesture.enabled = true
						self.untuckGesture.enabled = false
				})
		}
	}

	/** updateConstraints */
	override func updateConstraints() {
		removeAllConstraints()
		super.updateConstraints()
		stretchSubview(label)
	}

	/**
	handleSwipe:

	:param: gesture UISwipeGestureRecognizer
	*/
	func handleSwipe(gesture: UISwipeGestureRecognizer) {
		if gesture.state == .Ended {
			if gesture === tuckGesture { tuck() }
			else if gesture === untuckGesture { untuck() }
		}
	}

	/** attachTuckGestures */
	func attachTuckGestures() {
		let tuckGesture = MSSwipeGestureRecognizer(target: self, action: "handleSwipe:")
		tuckGesture.nametag = "'\(name)'-tuck"
		tuckGesture.enabled = false
		tuckGesture.direction = tuckDirection
		tuckGesture.quadrant = quadrant
		window?.addGestureRecognizer(tuckGesture)
		self.tuckGesture = tuckGesture

		let untuckGesture = MSSwipeGestureRecognizer(target: self, action: "handleSwipe:")
		untuckGesture.nametag = "'\(name)'-untuck"
		untuckGesture.enabled = false
		untuckGesture.direction = untuckDirection
		untuckGesture.quadrant = quadrant
		window?.addGestureRecognizer(untuckGesture)
		self.untuckGesture = untuckGesture
	}

	/** didMoveToWindow */
	override func didMoveToWindow() {
		if model.isPanel && !editing && window != nil { attachTuckGestures() }
		super.didMoveToWindow()
	}

	/**
	kvoRegistration

	:returns: [String:(MSKVOReceptionist) -> Void]
	*/
	override func kvoRegistration() -> [String:(MSKVOReceptionist) -> Void] {
		let registry = super.kvoRegistration()
		registry["label"] = {
			(receptionist: MSKVOReceptionist) -> Void in
				if let v = receptionist.observer as? ButtonGroupView {
					if let text = receptionist.change[NSKeyValueChangeNewKey] as? NSAttributedString { v.label.attributedText = text }
					else { v.label.attributedText = nil }
				}
		}
		return registry
	}

	/** initializeIVARs */
	override func initializeIVARs() {
		super.initializeIVARs()
		shrinkwrap = true
		resizable = true
		moveable = true
		if model.role & REButtonGroupRoleToolbar == REButtonGroupRoleToolbar {
			setContentCompressionResistancePriority(.Required, forAxis: .Horizontal)
			setContentCompressionResistancePriority(.Required, forAxis: .Vertical)
		}
	}

	/** didMoveToSuperview */
	override func didMoveToSuperview() {
		super.didMoveToSuperview()
		if superview != nil && model.isPanel && !editing {
			var attribute1 = NSLayoutAttribute.NotAnAttribute
			var attribute2 = attribute1
			switch model.panelLocation {
				case .Top:
					attribute1 = .Bottom
					attribute2 = .Top
					tuckDirection = .Up
					untuckDirection = .Down
					quadrant = .Up

				case .Bottom:
					attribute1 = .Top
					attribute2 = .Bottom
					tuckDirection = .Down
					untuckDirection = .Up
					quadrant = .Down

				case .Left:
					attribute1 = .Right
					attribute2 = .Left
					tuckDirection = .Left
					untuckDirection = .Right
					quadrant = .Left

				case .Right:
					attribute1 = .Left
					attribute2 = .Right
					tuckDirection = .Right
					untuckDirection = .Left
					quadrant = .Right

				default:
					break
			}

			let tuckedConstraint = NSLayoutConstraint(item: self,
				                                        attribute: attribute1,
				                                        relatedBy: .Equal,
				                                        toItem: superview,
				                                        attribute: attribute2,
				                                        multiplier: 1.0,
				                                        constant: 0.0)
			tuckedConstraint.priority = 999
			self.tuckedConstraint = tuckedConstraint

			let untuckedConstraint = NSLayoutConstraint(item: self,
				                                        attribute: attribute2,
				                                        relatedBy: .Equal,
				                                        toItem: superview,
				                                        attribute: attribute1,
				                                        multiplier: 1.0,
				                                        constant: 0.0)
			untuckedConstraint.priority = 1
			self.untuckedConstraint = untuckedConstraint

			superview.addConstraints([tuckedConstraint, untuckedConstraint])
		}
	}

	/**
	addSubelementView:

	:param: view RemoteElementView
	*/
	override func addSubelementView(view: RemoteElementView) {
		if locked {
			view.resizable = false
			view.moveable = false
		}
		if let buttonView = view as? ButtonView {
			if buttonView.role == REButtonRoleTuck {
				buttonView.tapAction = {self.tuck()}
			}
		}
		super.addSubelementView(view)
	}

	/** addInternalSubviews */
	override func addInternalSubviews() {
		super.addInternalSubviews()
		let label = UILabel.newForAutolayout()
		label.backgroundColor = UIColor.clearColor()
		addViewToContent(label)
		self.label = label
	}

	override var editingMode: REEditingMode {
		didSet {
			resizable = editingMode == .NotEditing
			moveable = editingMode == .NotEditing
			subelementInteractionEnabled = editingMode != .Remote
		}
	}

	/**
	intrinsicContentSize

	:returns: CGSize
	*/
	override func intrinsicContentSize() -> CGSize {
		if model.role == .Toolbar { return CGSize(width: UIScreen.mainScreen().bounds.width, height: 44.0) }
		else { return CGSize(square: UIViewNoIntrinsicMetric) }
	}

	/**
	buttonViewDidExecute:

	:param: buttonView ButtonView
	*/
	func buttonViewDidExecute(buttonView: ButtonView) {
		if model.autohide { dispatch_after(when: 1.0, queue: dispatch_get_main_queue()) { self.tuck() } }
	}

}
